datafusion/core/src/execution/session_state.rs

```diff
@@ -976,6 +837,482 @@ impl SessionState {
     }
 }
 
+/// A builder to be used for building [`SessionState`]'s. Defaults will be used for all values
+/// unless explicitly provided. Note that there is no `Default` or `new()` for SessionState,
+/// to avoid accidentally running queries or other operations without passing through
+/// the [`SessionConfig`] or [`RuntimeEnv`].
+pub struct SessionStateBuilder {
+    state: SessionState,
+    use_defaults: bool,
+}
+
+impl SessionStateBuilder {
+    /// Returns new [`SessionStateBuilder`] using the provided
+    /// [`SessionConfig`] and [`RuntimeEnv`].
+    pub fn new_with_config_rt(
+        config: SessionConfig,
+        runtime_env: Arc<RuntimeEnv>,
+    ) -> Self {
+        let session_id = Uuid::new_v4().to_string();
+        let catalog_list =
+            Arc::new(MemoryCatalogProviderList::new()) as Arc<dyn CatalogProviderList>;
+
+        Self {
+            state: SessionState {
+                session_id,
+                analyzer: Analyzer::new(),
+                expr_planners: vec![],
+                optimizer: Optimizer::new(),
+                physical_optimizers: PhysicalOptimizer::new(),
+                query_planner: Arc::new(DefaultQueryPlanner {}),
+                catalog_list,
+                table_functions: HashMap::new(),
+                scalar_functions: HashMap::new(),
+                aggregate_functions: HashMap::new(),
+                window_functions: HashMap::new(),
+                serializer_registry: Arc::new(EmptySerializerRegistry),
+                file_formats: HashMap::new(),
+                table_options: TableOptions::default_from_session_config(
+                    config.options(),
+                ),
+                config,
+                execution_props: ExecutionProps::new(),
+                table_factories: HashMap::new(),
+                runtime_env,
+                function_factory: None,
+            },
+            use_defaults: true,
+        }
+    }
+
+    /// Returns a new [SessionStateBuilder] based on an existing [SessionState]
+    /// The session id for the new builder will be set to a unique value; all
+    /// other fields will be cloned from what is set in the provided session state
+    pub fn new_from_existing(existing: &SessionState) -> Self {
```

Comment:
Since this `clone`s SessionState anyways, I think it would make more sense for this to take the argument by value

```suggestion
    pub fn new_from_existing(existing: SessionState) -> Self {
```

The rationale being there if the caller already has an owned SessionState they can use that and avoid a clone. This API basically requires them to do a clone

Reply:
In addition to this function, another nice to have UX feature would be a From impl, so users could do

```rust
let state: SessionState = get_existing_state();
let state = SessionStateBuilder::from(state)
  .with_optimizer_pass(Arc::new(MyNewOptimizer{}))
  .build()
```